File: nas-material-manager/ios/NASMaterialManager/NASMaterialManager/Data/Sync/SyncManager.swift
import Foundation

class SyncManager: ObservableObject {
    @Published var status = SyncStatus(state: .idle, lastSyncAt: nil, progress: nil)

    private let client: WebDAVClient
    private let materialRepo: MaterialRepository
    private let folderRepo: FolderRepository
    private var timer: Timer?
    private let pollInterval: TimeInterval = 45

    private var lastRemoteVersion: Date?
    private var localSyncVersion: Date?

    init(client: WebDAVClient, materialRepo: MaterialRepository, folderRepo: FolderRepository) {
        self.client = client
        self.materialRepo = materialRepo
        self.folderRepo = folderRepo
        loadLocalSyncVersion()
    }

    func startPolling() {
        stopPolling()
        timer = Timer.scheduledTimer(withTimeInterval: pollInterval, repeats: true) { [weak self] _ in
            Task { [weak self] in
                await self?.checkAndSync()
            }
        }
        Task {
            await checkAndSync()
        }
    }

    func stopPolling() {
        timer?.invalidate()
        timer = nil
    }

    func checkAndSync() async {
        await MainActor.run {
            status = SyncStatus(state: .checking, lastSyncAt: status.lastSyncAt, progress: nil)
        }

        do {
            let remoteVersion = try await fetchRemoteVersion()
            if localSyncVersion == nil || remoteVersion > localSyncVersion! {
                await downloadAndMergeChanges()
                localSyncVersion = remoteVersion
                saveLocalSyncVersion()
            }
            await MainActor.run {
                status = SyncStatus(state: .idle, lastSyncAt: Date(), progress: nil)
            }
        } catch {
            await MainActor.run {
                status = SyncStatus(state: .error(error.localizedDescription), lastSyncAt: status.lastSyncAt, progress: nil)
            }
        }
    }

    private func fetchRemoteVersion() async throws -> Date {
        do {
            let data = try await client.downloadFile(path: ".sync_version")
            if let dateString = String(data: data, encoding: .utf8), !dateString.isEmpty {
                return ISO8601DateFormatter().date(from: dateString) ?? Date.distantPast
            }
        } catch {
        }
        return Date.distantPast
    }

    private func downloadAndMergeChanges() async {
        await MainActor.run {
            status = SyncStatus(state: .downloading, lastSyncAt: status.lastSyncAt, progress: 0.0)
        }

        let indexFile = try? await client.downloadIndexFile(path: ".material_index.json")
        if let index = indexFile {
            mergeIndexFile(index, folderPath: "/")
        }

        await MainActor.run {
            status = SyncStatus(state: .downloading, lastSyncAt: status.lastSyncAt, progress: 1.0)
        }
    }

    private func mergeIndexFile(_ index: IndexFile, folderPath: String) {
        for (filename, fileIndex) in index.files {
            let path = folderPath.isEmpty ? filename : "\(folderPath)/\(filename)"
            let existingID = path.data(using: .utf8)?.base64EncodedString()
            // 简化实现：实际需要更完善的查找逻辑
        }
    }

    func uploadChanges(for materials: [Material], folderPath: String) async throws {
        await MainActor.run {
            status = SyncStatus(state: .uploading, lastSyncAt: status.lastSyncAt, progress: 0.0)
        }

        var filesDict: [String: MaterialIndex] = [:]
        for material in materials {
            let tagsIndex = TagsIndex(usage: material.tags.usage.rawValue, viral: material.tags.viral.rawValue)
            let fileIndex = MaterialIndex(
                title: material.title,
                description: material.description,
                tags: tagsIndex,
                updatedAt: material.localUpdatedAt,
                fileSize: material.fileSize,
                fileModifiedAt: material.fileModifiedAt
            )
            filesDict[material.filename] = fileIndex
        }

        let indexFile = IndexFile(
            version: 1,
            updatedAt: Date(),
            files: filesDict
        )

        try await client.uploadIndexFile(path: ".material_index.json", index: indexFile)
        try await touchSyncVersion()

        localSyncVersion = Date()
        saveLocalSyncVersion()

        await MainActor.run {
            status = SyncStatus(state: .idle, lastSyncAt: Date(), progress: 1.0)
        }
    }

    private func touchSyncVersion() async throws {
        let dateString = ISO8601DateFormatter().string(from: Date())
        let data = dateString.data(using: .utf8)!
        try await client.uploadFile(path: ".sync_version", data: data, contentType: "text/plain")
    }

    private func loadLocalSyncVersion() {
        if let dateString = UserDefaults.standard.string(forKey: "last_sync_version"),
           let date = ISO8601DateFormatter().date(from: dateString) {
            localSyncVersion = date
        }
    }

    private func saveLocalSyncVersion() {
        if let date = localSyncVersion {
            UserDefaults.standard.set(ISO8601DateFormatter().string(from: date), forKey: "last_sync_version")
        }
    }
}

File: nas-material-manager/flutter/lib/data/sync/sync_manager.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../webdav/webdav_client.dart';
import '../models/index_file.dart';
import '../models/material.dart';
import '../hive/hive_service.dart';
import 'sync_state.dart';

class SyncManager extends ChangeNotifier {
  SyncStatus _status = SyncStatus(state: SyncState.idle);
  SyncStatus get status => _status;

  final WebDAVClient client;
  final HiveService hiveService;
  Timer? _timer;
  static const _pollInterval = Duration(seconds: 45);

  DateTime? _localSyncVersion;

  SyncManager({required this.client, required this.hiveService}) {
    _loadLocalSyncVersion();
  }

  void startPolling() {
    stopPolling();
    _timer = Timer.periodic(_pollInterval, (_) => _checkAndSync());
    _checkAndSync();
  }

  void stopPolling() {
    _timer?.cancel();
    _timer = null;
  }

  Future<void> _checkAndSync() async {
    _updateStatus(SyncStatus(state: SyncState.checking, lastSyncAt: _status.lastSyncAt));

    try {
      final remoteVersion = await _fetchRemoteVersion();
      if (_localSyncVersion == null || remoteVersion.isAfter(_localSyncVersion!)) {
        await _downloadAndMergeChanges();
        _localSyncVersion = remoteVersion;
        await _saveLocalSyncVersion();
      }
      _updateStatus(SyncStatus(state: SyncState.idle, lastSyncAt: DateTime.now()));
    } catch (e) {
      _updateStatus(SyncStatus(
        state: SyncState.error,
        lastSyncAt: _status.lastSyncAt,
        errorMessage: e.toString(),
      ));
    }
  }

  Future<DateTime> _fetchRemoteVersion() async {
    try {
      final data = await client.downloadFile('.sync_version');
      final dateString = utf8.decode(data);
      if (dateString.isNotEmpty) {
        return DateTime.parse(dateString);
      }
    } catch (e) {
    }
    return DateTime.fromMicrosecondsSinceEpoch(0);
  }

  Future<void> _downloadAndMergeChanges() async {
    _updateStatus(SyncStatus(
      state: SyncState.downloading,
      lastSyncAt: _status.lastSyncAt,
      progress: 0.0,
    ));

    final indexFile = await client.downloadIndexFile('.material_index.json');
    if (indexFile != null) {
      await _mergeIndexFile(indexFile, '/');
    }

    _updateStatus(SyncStatus(
      state: SyncState.downloading,
      lastSyncAt: _status.lastSyncAt,
      progress: 1.0,
    ));
  }

  Future<void> _mergeIndexFile(IndexFile index, String folderPath) async {
    for (final entry in index.files.entries) {
      final filename = entry.key;
      final fileIndex = entry.value;
      final path = folderPath.isEmpty ? filename : '$folderPath/$filename';

      final existing = hiveService.getMaterial(path);
      if (existing != null) {
        if (fileIndex.updatedAt.isAfter(existing.localUpdatedAt)) {
          final updated = existing.copyWith(
            title: fileIndex.title,
            description: fileIndex.description,
            tags: existing.tags.copyWith(
              usage: UsageTag.fromRawValue(fileIndex.tags.usage),
              viral: ViralTag.fromRawValue(fileIndex.tags.viral),
            ),
            localUpdatedAt: fileIndex.updatedAt,
          );
          await hiveService.updateMaterial(updated);
        }
      } else {
        final material = Material(
          filename: filename,
          path: path,
          title: fileIndex.title,
          description: fileIndex.description,
          tags: MaterialTags(
            usage: UsageTag.fromRawValue(fileIndex.tags.usage),
            viral: ViralTag.fromRawValue(fileIndex.tags.viral),
          ),
          fileSize: fileIndex.fileSize,
          fileModifiedAt: fileIndex.fileModifiedAt,
          localUpdatedAt: fileIndex.updatedAt,
        );
        await hiveService.saveMaterial(material);
      }
    }
    notifyListeners();
  }

  Future<void> uploadChanges(List<Material> materials, String folderPath) async {
    _updateStatus(SyncStatus(
      state: SyncState.uploading,
      lastSyncAt: _status.lastSyncAt,
      progress: 0.0,
    ));

    final filesDict = <String, MaterialIndex>{};
    for (final material in materials) {
      final tagsIndex = TagsIndex(
        usage: material.tags.usage.rawValue,
        viral: material.tags.viral.rawValue,
      );
      final fileIndex = MaterialIndex(
        title: material.title,
        description: material.description,
        tags: tagsIndex,
        updatedAt: material.localUpdatedAt,
        fileSize: material.fileSize,
        fileModifiedAt: material.fileModifiedAt,
      );
      filesDict[material.filename] = fileIndex;
    }

    final indexFile = IndexFile(
      version: 1,
      updatedAt: DateTime.now().toUtc(),
      files: filesDict,
    );

    await client.uploadIndexFile('.material_index.json', indexFile);
    await _touchSyncVersion();

    _localSyncVersion = DateTime.now().toUtc();
    await _saveLocalSyncVersion();

    _updateStatus(SyncStatus(
      state: SyncState.idle,
      lastSyncAt: DateTime.now(),
      progress: 1.0,
    ));
  }

  Future<void> _touchSyncVersion() async {
    final dateString = DateTime.now().toUtc().toIso8601String();
    final data = utf8.encode(dateString);
    await client.uploadFile('.sync_version', data, contentType: 'text/plain');
  }

  Future<void> _loadLocalSyncVersion() async {
    final prefs = await SharedPreferences.getInstance();
    final dateString = prefs.getString('last_sync_version');
    if (dateString != null) {
      _localSyncVersion = DateTime.tryParse(dateString);
    }
  }

  Future<void> _saveLocalSyncVersion() async {
    final prefs = await SharedPreferences.getInstance();
    if (_localSyncVersion != null) {
      await prefs.setString('last_sync_version', _localSyncVersion!.toIso8601String());
    }
  }

  void _updateStatus(SyncStatus newStatus) {
    _status = newStatus;
    notifyListeners();
  }
}
